File: src/csharp/Foundgine.Core/Semantic/Planning/Mutation/UpsertIntent.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Planning.Mutation;

/// <summary>
/// Provider-neutral upsert intent. ConflictColumns identify the logical
/// uniqueness key; when omitted the entity primary key is used.
/// ReturnFields are semantic fields requested from the resulting row.
/// </summary>
public sealed record UpsertIntent(
    EntityId Entity,
    IReadOnlyList<MutationFieldValue> Fields,
    IReadOnlyList<ColumnId>? ConflictColumns = null,
    IReadOnlyList<FieldId>? ReturnFields = null) : IMutationIntent
{
    public MutationKind Kind => MutationKind.Upsert;
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/mutation/UpsertIntent.java
package com.foundgine.core.semantic.planning.mutation;

import com.foundgine.core.abstractions.*;
import java.util.*;

public record UpsertIntent(EntityId entity, List<MutationFieldValue> fields, List<ColumnId> conflictColumns,
		List<FieldId> returnFields) implements IMutationIntent {
	public UpsertIntent {
		Objects.requireNonNull(entity);
		fields = fields == null ? List.of() : List.copyOf(fields);
		conflictColumns = conflictColumns == null ? null : List.copyOf(conflictColumns);
		returnFields = returnFields == null ? null : List.copyOf(returnFields);
	}

	public MutationKind kind() {
		return MutationKind.UPSERT;
	}

	public UpsertIntent(EntityId entity, List<MutationFieldValue> fields) {
		this(entity, fields, null, null);
	}
}
